packages/eslint-plugin/tests/rules/no-unnecessary-type-parameters.test.ts

```diff
@@ -16,11 +16,6 @@ const ruleTester = new RuleTester({
 
 ruleTester.run('no-unnecessary-type-parameters', rule, {
   valid: [
-    `
-      class ClassyArray<T> {
-        arr: T[];
-      }
-    `,
```

Comment:
Err, shouldn't this still be allowed? `ClassyArray` is fine per this rule.

```ts
const classy = new ClassyArray<number>();
classy.arr = [1, 2, 3];
```

Reply:
"`ClassyArray` should be invalid, because why not just use an array?" Cit. @danvk.

I agree by the way.

Reply:
Gotcha. I don't think we got a consensus agreement on allowing `ClassyArray`. `ClassyArray` is, roughly speaking, code that someone might reasonably write. Expanding on https://github.com/typescript-eslint/typescript-eslint/issues/9529#issuecomment-2221917930:

```ts
class LabeledArray<T> {
  arr: T[] | undefined;
  label: string;

  constructor(label: string) {
    this.arr = [];
    this.label = label;
  }
}

const labeledArray = new LabeledArray<string>("letters");
labeledArray.arr.at(0); // string | undefined
```

I think we have consensus on two things:

* Per @danvk's https://github.com/typescript-eslint/typescript-eslint/issues/9529#issuecomment-2223224801, making _truly safe_ things as safe:
  * Treating tuple types as singular
  * Treating array types in a parameter slot (not return type) as singular
* Adding an _opt-in option_ for things like arrays.

`ClassyArray`/`LabeledArray` wouldn't qualify for _"array types in a parameter slot"_.

Reply:
@JoshuaKGoldberg I added some complexity, but now your example isn't considered an error anymore, while the following are at the moment: 

```ts
class Joiner<T extends string | number> {
  join: (els: T[]) => string;
}

class Joiner<T extends string | number> {
  join(els: T[]) {
    return els.map(el => '' + el).join(',');
  }
}

class Joiner<T extends string | number> {
  join = (els: T[]) => {
    return els.map(el => '' + el).join(',');
  };
}
```

Reply:
Should

```ts
class ClassyArray<T> {
        arr: [T];
        label: string;
}
```

be allowed?

Reply:
I find @JoshuaKGoldberg's `LabeledArray` example persuasive. Let's allow it.

`Joiner` should be disallowed since the `T[]` only appears in a parameter slot.

`ClassyTuple` should be allowed. I don't think the tuple-ness is really relevant to it, because shouldn't this also be allowed?

```ts
class LabeledBox<T> {
    value: T;
    label: string;
}
```

Reply:
I see the point about re-usability, but if we accept that as a rationale then I'm not sure we'd have any invalid cases left thanks to [instantiation expressions](https://devblogs.microsoft.com/typescript/announcing-typescript-4-7/#instantiation-expressions):

```ts
function getLength<T>(array: T[]) {
  return array.length;
}

const getStringArrayLength = getLength<string>;
const getLogLevelArrayLength = getLength<"info" | "warn">;
```

Reply:
Yeah, why should the first be allowed by this rule while the second one is not?

```ts
namespace WithClass {

    declare class LeveledLogger<Level extends string> {
        log(level: Level, ...args: string[]): void;
    }

    const logger = new LeveledLogger<"info" | "warn">();

    logger.log("info", "a"); // Ok
    logger.log("warn", "a"); // Ok

    logger.log("error", "a"); // Not ok

}

namespace WithFunction {
    declare function leveledLogger<Level extends string>
        (level: Level, ...args: string[]): void

    const logger = leveledLogger<"info" | "warn">;

    logger("info", "a"); // Ok
    logger("warn", "a"); // Ok

    logger("error", "a"); // Not ok
}
```

Reply:
_Blurgh_ instantiation expressions! I'd forgotten about them. I've never seen someone use them in the wild.

The rule is essentially useless if it considers instantiation expressions a potential multiple use of parameters. That would invalidate roughly all functions! I think there's no reasonable way to support them and this rule. They'd probably be better as something explicitly mentioned in the rule's docs (#9680). cc @kirkwaiblinger 

In my mind, the difference between classes and function instantiation expressions is that classes are _made_ to be reused. I've written code like `LeveledLogger` before and seen others opt into it. But I can see an argument that it's kid of niche. What do you all think? Also I'd want to hear from @typescript-eslint/triage-team

Reply:
This is brain-melty stuff. 🫠 I've been thinking about this for far too long and not come up any concrete opinion yet 😆 

One question I have is - is there any difference in behavior between an instantiation expression and providing explicit type arguments? I see that they're represented differently in the AST, but, to my mind, 
```ts
const logger = new LeveledLogger<"info" | "warn">();
```
is the same as
```ts
const logger = new (LeveledLogger<"info" | "warn">)();
```
right?

When we directly instantiate the class with type arguments (or call a function with type arguments, like `f<T>()`) we're doing exactly the same thing as if we aliased the instantiation expression first as in `const ft = f<T>; ft();`, right?

Reply:
The reason I ask is that my _very very tentative_ instinct is to say that the current approach in the rule is basically incompatible with classes/functions intended to be used with instantiation expressions or with explicit type arguments.

Reply:
> to my mind, ... has no observable differences to ...

👍 agreement from me.

> current approach in the rule is basically incompatible with classes/functions intended to be used with instantiation expressions or with explicit type arguments

👍 agreement with me. For those cases, I don't see a way to make the rule actually useful at all.

Reply:
Cool.

So what gives me pause is that the logger case is _potentially useful_ and _doesn't introduce type unsafety_ because the type is used in parameter position on a method, rather than return type position. So, yes, the class could be written without, but it doesn't seem to incur any of the negatives that the rule kind of intends to lint against. 🤔 I'm having a hard time thinking about whether this analysis can be generalized into something we can lint for or whether it should be handled by users on a case by case basis. 

Similarly, I don't see how 
```ts
function f<T>(x: T): void {
    // any implementation
}
```
could _ever_ incur type unsafety. But it does in theory have niche use in instantiation expressions/explicit type argument scenarios. So - do we want to allow this (possibly via an option?)? Or just document it as a limitation?

Reply:
The trouble with the type unsafety analysis is - I'm not sure that you can incur type unsafety without doing something else that is also problematic. The canonical example, 
```ts
function totallyUnsafe<T>(): T {
    return JSON.parse('"anything"');
}
```
can be defended against with `no-unsafe-return`. And, similarly, anything that uses `as T` is overtly unsafe regardless of the single-use generic. 

So, the question in my mind is - does this rule _really_, _really_, matter for type safety reasons? Or is it a purely stylistic rule? 

I think deciding on an answer to that question may inform the judgement calls we'll make here and elsewhere.

Reply:
I see it as for both stylistic and type safety. The removal of unnecessary constructs (`function log<T>(value: T): void`) is useful, as is hiding assertions (`function parse<T>(): T`).

> whether this analysis can be generalized into something we can lint for or whether it should be handled by users on a case by case basis.

So it sounds like, for now, the conclusion is _no_: that's just not a pattern compatible with this rule. So that'd be a good FAQ. Am I misreading the conclusion?

Reply:
👋 following up @jfet97, @kirkwaiblinger and I chatted and we're still on this opinion.  That in _this_ PR the only change to be made is special casing tuples and parameter arrays as single-use.